Make sessions obey geometry

bobwhitelock opened the pull request:
This PR is sessions support for https://github.com/alces-software/clusterware/pull/154; it consists of various changes to some session scripts to make them obey the geometry parameter which can now be set when starting a session, and is passed into the environment when the VNC server is started. These changes consist of:
- gnome, cinnamon - adapting the geometry variable read to fix the session geometry to be consistent with the one set in Clusterware;
- xfce - applying the same fix as for gnome and cinnamon sessions;
- chrome - setting the chrome window to match the dimensions of the session window; another fix to allow having multiple chrome sessions running is also included;
- default, terminal - passing the geometry through to the xterm's `-geometry` parameter;
- fvwm, trinity - no changes needed.

There are also a couple of things I noticed which I don't know if we want to handle, now or later, or are fine as-is:
- for the gnome/cinnamon/xfce clusterware-geometry fix, if a geometry is provided that is not in the list given by `xrandr` then the desktop will be resized to the largest value that is provided by `xrandr`;
- if a gnome or cinnamon session has been run on a cluster prior to it pulling this fix, then the updates in this PR will not be picked up as the geometry script is cached in `~/.local/share/clusterware/bin/clusterware-geometry.sh`; this will need to be removed for the update to take effect;
- there is still duplication between the geometry fix for gnome and cinnamon sessions, and now xfce sessions as well.

bobwhitelock committed:
Rename geometry var for consistency with CW

bobwhitelock committed:
Duplicate geometry code for XFCE

Like Gnome and Cinnamon, XFCE also wasn't obeying the VNC server's
geometry setting. Duplicating the fix for Gnome and Cinnamon solves
this, though we should probably factor the duplication out at some
point.

bobwhitelock committed:
Make Chrome window fill Chrome session on start

Previously Chrome would start at the size it last had at if it
had been opened previously, and it was not possible to resize the window
as the Chrome session type does not have a window manager.

This is particularly needed now that Clusterware users can specify a
session geometry on the command line, but could still have caused issues
previously if Chrome was opened and resized within a session of another
type and then a Chrome session was started.

bobwhitelock committed:
Allow running multiple Chrome sessions

bobwhitelock committed:
Set xterm geometry for terminal/default sessions

Set geometry of xterm to match geometry of VNC session for
terminal/default sessions, so the xterm fills the session rather than
having a black border in the rest of the session.

bobwhitelock committed:
Revert "Set xterm geometry for terminal/default sessions"

This reverts commit bf1d3a9f91409dfb0b9a114f2dba0d64792c8b91.

This does not actually work as intended, as xterm's `-geometry` flag
expects a geometry to be given as <columns>x<rows> rather than by width
and height. We'll need to put more thought into how we want to handle
different geometries for these session types.

mjtko commented on xfce/session.sh:
```diff
@@ -19,6 +20,31 @@ destdir="$(xdg_config_home)/xfce4/xfconf/xfce-perchannel-xml"
+      destdir="$(xdg_data_home)/clusterware/bin"
+      geom_sh="${destdir}/${f}"
+      mkdir -p "${destdir}"
+      cp "${cw_ROOT}"/etc/sessions/cinnamon/${f} "${geom_sh}"
```
I think this (and l39 below) should be `sessions/xfce`.

bobwhitelock replied:
I think you're right, I'll change it.

bobwhitelock committed:
Correct cinnamon -> xfce in xfce session

mjtko commented:
> if a geometry is provided that is not in the list given by xrandr then the desktop will be resized to the largest value

The behaviour here seems to be "desktop runs at default for the chosen session".  We can live with this. :)

> if a gnome or cinnamon session has been run on a cluster prior to it pulling this fix, then the updates in this PR will not be picked up

This is primarily intended for ephemeral clusters, so that's ok.

> there is still duplication

Not worried about this.

mjtko commented on xfce/session.sh:
```diff
@@ -19,6 +20,31 @@ destdir="$(xdg_config_home)/xfce4/xfconf/xfce-perchannel-xml"
+  fi
+}
+
+if distro_is el7; then
```
I would imagine that XFCE behaviour is the same between EL7 and EL6.  In other places behaviour does vary (GNOME in EL6 is a different version from GNOME in EL7).

It's not a big deal but probably worth dropping this condition IMO.

bobwhitelock replied:
Sure, I've removed it.

bobwhitelock committed:
Always install geometry script for XFCE

As suggested by @mjtko, as XFCE does not differ much between EL6 and EL7.

mjtko commented:
Actually, regarding "geometry that is not in list", how about we do this:

```
current_geometry=$(xrandr -q | awk -F 'current' -F',' 'NR==1 {gsub("( |current)",""); print $2}')
desired="${cw_SESSION_geometry:-1024x768}"
if [ "$current_geometry" != "$desired" ] ; then
    if ! xrandr -q | tail -n+3 | grep -q "$desired"; then
      desired="1024x768"
    fi
    xrandr --output VNC-0 --mode $desired
fi
```

Probably better than going massive?

bobwhitelock commented:
That looks reasonable to me - have you checked this works fine/ do you want me to add this in or do you want to push it?

mjtko commented:
Yup, I've checked it works for `gnome`. I don't see any reason why it wouldn't work for the other sessions -- if you have them to hand though, it would be handy to give it a once over with `xfce` and `cinnamon` too.

bobwhitelock committed:
Fix for sessions defaulting to largest size

This would happen for gnome/cinnamon/xfce sessions when a geometry not
in the list given by xrandr was given. Thanks @mjtko for the fix (ref
https://github.com/alces-software/clusterware-sessions/pull/9#issuecomment-219998687.

bobwhitelock commented:
@mjtko - I've checked your fix and it appears to work fine for cinnamon/xfce - though weirdly xfce now does appear to be obeying the geometry parameter from the VNC session, though it definitely wasn't before - not sure why, possibly it's interacting with something else?

mjtko commented:
Possibly that's cos you have multiple sessions available and XFCE is picking up the autostart script from one of the others? Dunno! Let's go with this anyway. 😄
